### ISSUE TITLE
[Feature] Enabling `tide-mode` workflow with TS *monorepo* (React)

### ISSUE THREAD

**Issue by david-alvarez-rosa** [ananthakumaran/tide#472]
Hi Emacs wizards -- wanted to ask about how can work in a *monorepo*, were multiple teams have each their own package (repo).

Let `pkg-1`, ..., `pkg-n` be the packages corresponding to `team-1`, ..., `team-n`, respectively.

At compile-time, those directories are (symbolically) linked to `pkg-main`

```text
pkg-main/
    .gitignore
    tsconfig.json
    packages/
        pkg-1 -> ../pkg-1
        ...
        pkg-n -> ../pkg-n
 pkg-1/
    .gitignore
 ...
 pkg-n/
    .gitignore
```

Any tips on how can this be handled by `tide-mode`?

--

_PS_  Why not enabling repo sponsorship? I'd be happy to contribute :)

**Comment by ananthakumaran** [comment]
When you open any .ts file, tide walks up the directory till it finds a directory with tsconfig.json file and it will start tsserver with that directory as the project root. https://github.com/ananthakumaran/tide/blob/master/tide.el#L323 You can override the logic by setting the `tide-project-root` buffer local variable. I am not sure whether you want tide to work at individual project level or whole repo level. Individual project level is usually straightforward, just need to have a tsconfig at the project root directory.

**Comment by david-alvarez-rosa** [comment]
Thanks @ananthakumaran —- unfortunately for me this doesn't really work as expected with `projectile`.

**Comment by forty** [comment]
What I do is to force `tide-project-root` to be the project root and I fill typescript project references in my tsconfig.json. This way I have a single tsserver managing my whole monorepo, works like a charm.
I use pnpm to manage the monorepo and meta-updater to maintain the typescript project references.

The only thing that doesn't work is tide-project-errors are the response from tsserver seems to big to process, but I'm not surprised, it's a really large monorepo.

**Comment by wu-lee** [comment]
This may warrant a separate issue, but whilst I remember: I hit this problem. Specifically: in tide mode, within a workspace of a monorepo, tide's "intellisense" was advising me of errors for the wrong version of Typescript.

For example, `"some string".replaceAll(blah)` was highlit as an error, because an older node was inferred for which `replaceAll` was not yet defined.

Using `ps aux | grep tsserver` I could see that the typescript server was the generic one supplied by tide, and not the one appropriate for this workspace (which targetted `node22`). And that seemed to be because the `node_modules` folder was absent, as if I symlinked it to the project root's `node_modules` things behaved correctly.

But it's now normal for there to be no `node_modules` directory in a workspace, if there are no dependencies specific to it. So something else must be missing.

I think I've tracked it down to the tide heuristics for finding the tsserver to use.  It uses projectile to find the project root. However, although projectile is present in my setup, it was not loaded by default. Loading projectile with `(require 'projectile)` prior to setting up tide in init.el fixes that problem for me.

*To summarise: perhaps the tide docs should mention that `(require 'projectile)` is needed for correct operation in this case?*

Let me know if you'd like me to submit an issue for this...

tide version: melpa 20241019.2101
emacs version: 28.2
